### ISSUE TITLE
how to use @snoopi

### ISSUE THREAD

**Issue by maartenvd** [timholy/SnoopCompile.jl#25]
Sorry if this is not the right place to post usage questions, but I'm really excited to try out the new snoopi macro (because inference is what I'm having problems with). However, when I try to snoopi on a call that takes 1+seconds to infer, I get the following trace:

```
 (4.100799560546875e-5, MethodInstance for eltype(::Type{Array{Expr,2}}))                                                                                                                                                                                         
...                                    
 (0.0013608932495117188, MethodInstance for methods(::Any, ::Any))                                                                                                                                                                                                
 (1.1331579685211182, MethodInstance for *(::StridedView{Float64,2,Array{Float64,1},typeof(identity)}, ::StridedView{Float64,2,Array{Float64,1},typeof(identity)}))   
```

The call I'm curious about is the multiplication, but also in how long all methods called by multiplication take. The problem is that the sum of all timings given by snoopi does not equal the time it takes to infer the multiplication call. There seems to be function calls missing, which perhaps got inlined, but even with --inline=no there is no change.

to be clear, I ran:

```
using SnoopCompile,Stridedc
a = StridedView(rand(5,5))
vs = @snoopi a*a
```

**Comment by timholy** [comment]
Usage questions are fine. It reports only the "top-level" method; if `foo(::Int)` calls `bar(::Int)`, `foo` cannot be inferred without first inferring `bar`. But only the inference of `foo` gets reported.

It may help to look at how [it works](https://github.com/timholy/SnoopCompile.jl/blob/92457419c94f1a484cb1a9fbc9df2e0fa2f47a82/src/SnoopCompile.jl#L19-L21): there's nothing to intercept recursive calls, only the entry point.

If you can turn your confusion into improved documentation via a pull request, I'd be very grateful!

**Comment by maartenvd** [comment]
Oh, I assumed that by calling 
`jl_set_typeinf_func`
inference would from now on use your supplied function to do everything, so therefore I expected to see a recursive trace.

I can take a look at the documentation :)

**Comment by timholy** [comment]
> Oh, I assumed that by calling jl_set_typeinf_func inference would from now on use your supplied function to do everything

It does, but it only intercepts calls to `typeinf_ext`. That's the "entry point" from Julia's C side; the C code doesn't handle the recursion, that's all done by the Julia side from `typeinf_ext`. To intercept recursive calls we'd have to intercept `abstract_call` as well, and that involves more rewriting/overwriting the methods in the compiler. This is the only place I've found that you can observe without substantial disruption.

**Comment by timholy** [comment]
Fixed by #35
